src/collapse/xctrace.rs

```diff
@@ -0,0 +1,492 @@
+            <frame id="29" name="std::rt::lang_start_internal::hfc27b745d167a74d" addr="0x102b0a324"></frame>
+            <frame id="30" name="main" addr="0x102af5fa0"></frame>
+            <frame id="31" name="start" addr="0x18d373e50"></frame>
+        </backtrace>
+    </row>
+    <row>
+        <backtrace id="39">
+            <frame id="40" name="rust_test2::foo::ha31fba0d06a8a3eb" addr="0x102af5d95"></frame>
+            <frame ref="24"/>
+            <frame ref="25"/>
+            <frame ref="27"/>
+            <frame ref="29"/>
+            <frame ref="30"/>
+            <frame ref="31"/>
+        </backtrace></row>
+</node>
+</trace-query-result>
+```
+ */
+
+// ----------- attribute names -----------
+
+/// Reference to a previous declared backtrace/frame, it's value is id.
+const REF: &[u8] = b"ref";
+/// Id of a backtrace/frame
+const ID: &[u8] = b"id";
+/// Symbolicated name of a frame
+const NAME: &[u8] = b"name";
+
+// -----------    tag names    -----------
+
+/// Root tag of xctrace's xml output
+const TRACE_QUERY_RESULT: &[u8] = b"trace-query-result";
+/// Container of a sample set
+const NODE: &[u8] = b"node";
+/// Container of a sample.
+const ROW: &[u8] = b"row";
+/// Stack backtrace
+const BACKTRACE: &[u8] = b"backtrace";
+/// Stack frame
+const FRAME: &[u8] = b"frame";
+
+// Is this a tag we are interested in?
+fn is_interested_tag(tag: &[u8]) -> bool {
+    matches!(tag, TRACE_QUERY_RESULT | NODE | ROW | BACKTRACE | FRAME)
+}
+
+/// Context of collapsing a xctrace's `Time Profiler` xml
+#[derive(Default)]
+struct CollapseContext {
+    /// xml tag backtrace
+    state_backtrace: TagBacktrace,
+    // --------- per-xml caches below -----------
+    /// backtrace_id <--> BackTrace
+    backtraces: BTreeMap<u64, Arc<Backtrace>>,
+    /// backtrace_id <--> Frame
+    frames: BTreeMap<u64, Arc<Frame>>,
+}
+
+// Note that sometimes same Backtrace have different id, we need to merge them
+// before writing folded file.
+struct BacktraceOccurrences {
+    /// How many times the backtrace occurred.
+    num: u64,
+    /// Backtrace content
+    backtrace: Arc<Backtrace>,
+}
+
+/// Backtrace of scanned xml tags.
+#[derive(Default)]
+struct TagBacktrace {
+    backtrace: Vec<CurrentTag>,
+}
+
+impl TagBacktrace {
+    fn push_back(&mut self, tag: CurrentTag) {
+        self.backtrace.push(tag);
+    }
+
+    /// If `name` matches the top tag of the stack, the tag is popped.
+    fn pop_with_name(&mut self, name: &[u8]) -> Option<CurrentTag> {
+        self.backtrace
+            .last()
+            .is_some_and(|t| t.matches(name))
+            .then(|| self.backtrace.pop())?
+    }
+
+    fn top_mut(&mut self) -> Option<&mut CurrentTag> {
+        self.backtrace.last_mut()
+    }
+}
+
+/// The tag we are scanning, with additional states.
+enum CurrentTag {
+    TraceQueryResult { nodes: Vec<Node> },
+    Node { rows: Vec<Row> },
+    Row { backtrace: Option<Arc<Backtrace>> },
+    Backtrace { id: u64, frames: Vec<Arc<Frame>> },
```

Comment:
Do we actually need the `Arc`s here? My sense is that we can just pass around backtrace/frame IDs everywhere, and look them up when we need their contents?

Reply:
Fixed.